rpc: optimize VerifyClockOffset a bit

stevendanna opened the pull request:
This replaces some of our stats calculations with new versions that assume sorted input, avoiding a good deal of the copying and sorting contained in library we were using.

Before:

    BenchmarkVerifyClockOffset  5313    203408 ns/op    82118 B/op    5 allocs/op

After:

    BenchmarkVerifyClockOffset 10000    102450 ns/op    16489 B/op    1 allocs/op

Fixes #147825

Release note: None

stevendanna committed:
rpc: optimize VerifyClockOffset a bit

This replaces some of our stats calculations with new versions that
assume sorted input, avoiding a good deal of the copying and sorting
contained in library we were using.

Before:

    BenchmarkVerifyClockOffset  5313    203408 ns/op    82118 B/op    5 allocs/op

After:

    BenchmarkVerifyClockOffset 10000    102450 ns/op    16489 B/op    1 allocs/op

Fixes #147825

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147827)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a suggestion to avoid the remaining sort.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-:-OSPFvjo4OKi124-GNkQ:b-f4njvr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/rpc/clock_offset.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPDjaS1WijFNqF7Wz3:-OSPDjaS1WijFNqF7Wz4:b67weu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L349)):*
> ```Go
> 		defer r.mu.Unlock()
> 		// Each measurement is recorded as its minimum and maximum value.
> 		offs := make(stats.Float64Data, 0, 2*len(r.mu.offsets))
> ```

Can we store this slice in `RemoteClockMonitor`? We'd have to hold the lock for the entire function.
___
*[`pkg/rpc/clock_offset.go` line 371 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPDBzv-8IjVYLQutVl:-OSPDBzw8KcyK1GRerqC:brl1nyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L371)):*
> ```Go
> 	stdDev := StandardDeviationPopulationKnownMean(offsets, mean)
> 	median := MedianSortedInput(offsets)
> 	medianAbsoluteDeviation := MedianAbsoluteDeviationPopulationSortedInputMutatesInput(offsets)
> ```

Add a comment here with a warning that offsets is being "destroyed" here. (I know it's in the function name but it's easily missed)
___
*[`pkg/rpc/clock_offset.go` line 502 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPCMJk81F1xzg5PeYR:-OSPCMJk81F1xzg5PeYS:bz9fao4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L502)):*
> ```Go
> // sorted.
> func MedianSortedInput(sortedInput stats.Float64Data) float64 {
> 	l := len(sortedInput)
> ```

[nit] I'd add a CrdbTestBuild check that verifies that it's sorted
___
*[`pkg/rpc/clock_offset.go` line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPBwvf2lyjbZ784rUp:-OSPBwvf2lyjbZ784rUq:bzidsip) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L506)):*
> ```Go
> 		return math.NaN()
> 	} else if l == 1 {
> 		return sortedInput[0]
> ```

[nit] this case isn't necessary (`sortedInput[l/2]` works just fine)
___
*[`pkg/rpc/clock_offset.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPEnDo2WQjg5fycrHW:-OSPEnDp5SL55MrQg-jV:b-ke9jxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L525)):*
> ```Go
> 	}
> 
> 	m := MedianSortedInput(sortedInput)
> ```

We can do better here. We can peel off the largest deviation n/2 times.  The largest deviation is always at the beginning or the end since the data is sorted:

```
a := sortedInput
for len(a) > len(sortedInput)/2  /* possible off by one in this condition */ {
  // Remove largest deviation.
  if m - a[0] >= a[len(a)-1] - m {
    a = a[1:]
  } else {
    a = a[:len(a)-1]
  }
}
return either max(m-a[0], a[len(a)-1]-m) or their average
```

This avoids the sort and the mutation.


<!-- Sent from Reviewable.io -->

stevendanna committed:
rpc: remove another sort from VerifyClockOffsets

We can calculate the median absolute deviation without re-sorting. The
code is a bit more complicated, but it does seem to be a win for the
benchmark.

```
goos: darwin
goarch: arm64
cpu: Apple M1 Pro
                  │  before.txt  │              after.txt               │
                  │    sec/op    │    sec/op     vs base                │
VerifyClockOffset   101.90µ ± 2%   84.69µ ± 20%  -16.88% (p=0.009 n=20)

                  │  before.txt  │              after.txt              │
                  │     B/op     │     B/op      vs base               │
VerifyClockOffset   16.10Ki ± 0%   16.07Ki ± 0%  -0.19% (p=0.000 n=20)

                  │ before.txt │           after.txt            │
                  │ allocs/op  │ allocs/op   vs base            │
VerifyClockOffset   1.000 ± 0%   1.000 ± 0%  ~ (p=1.000 n=20) ¹
¹ all samples are equal
```

Epic: none
Release note: None

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147827)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[`pkg/rpc/clock_offset.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPEnDo2WQjg5fycrHW:-OSZGE7E6k-b-44R_pAm:b-r9pupq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L525)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can do better here. We can peel off the largest deviation n/2 times.  The largest deviation is always at the beginning or the end since the data is sorted:

```
a := sortedInput
for len(a) > len(sortedInput)/2  /* possible off by one in this condition */ {
  // Remove largest deviation.
  if m - a[0] >= a[len(a)-1] - m {
    a = a[1:]
  } else {
    a = a[:len(a)-1]
  }
}
return either max(m-a[0], a[len(a)-1]-m) or their average
```

This avoids the sort and the mutation.
</blockquote></details>

Thanks for the nudge here. I had tried this originally, it failed the tests, so I figured I had just mis-understood something.  I took another look based on your comment, and I think that this general idea works, it is just that in the case of an even length input, you have to look across for values for the two values needed to average.

This makes the code a bit more complicated, but does seem to be a win for the benchmark.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147827)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna)

___
*[`pkg/rpc/clock_offset.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPEnDo2WQjg5fycrHW:-OSZetwTEQUlitrlji97:bx1bney) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L525)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for the nudge here. I had tried this originally, it failed the tests, so I figured I had just mis-understood something.  I took another look based on your comment, and I think that this general idea works, it is just that in the case of an even length input, you have to look across for values for the two values needed to average.

This makes the code a bit more complicated, but does seem to be a win for the benchmark.
</blockquote></details>

Ah, indeed. You can remember the last value from the loop, something like this:
```
	last := 0
	for len(a) > len(sortedInput)/2 {
		// Remove largest deviation.
		if m-a[0] >= a[len(a)-1]-m {
			last = m - a[0]
			a = a[1:]
		} else {
			last = a[len(a)-1] - m
			a = a[:len(a)-1]
		}
	}
	if len(sortedInput)%2 == 1 {
		return float64(last)
	}
	return 0.5 * float64(last+max(m-a[0], a[len(a)-1]-m))

```


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147827)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[`pkg/rpc/clock_offset.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/147827#-OSPEnDo2WQjg5fycrHW:-OSaN6IG6uqrjOWfxLRG:b-y4k9cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab0e37522aa582a090ce2177f9b0e270acabbdc/pkg/rpc/clock_offset.go#L525)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, indeed. You can remember the last value from the loop, something like this:
```
	last := 0
	for len(a) > len(sortedInput)/2 {
		// Remove largest deviation.
		if m-a[0] >= a[len(a)-1]-m {
			last = m - a[0]
			a = a[1:]
		} else {
			last = a[len(a)-1] - m
			a = a[:len(a)-1]
		}
	}
	if len(sortedInput)%2 == 1 {
		return float64(last)
	}
	return 0.5 * float64(last+max(m-a[0], a[len(a)-1]-m))

```
</blockquote></details>

Ah of course. Thanks! We are down to 1 sort and 1 copy.  I'll merge without the change to move offs into the struct only because that changes the locking footprint and I know this lock has been heavily contended in the past.


<!-- Sent from Reviewable.io -->
